ui/tests/unit/abilities/job-test.js

```diff
@@ -0,0 +1,183 @@
+  });
+
+  test('it blocks job run for client tokens with a policy that has no submit-job capability', function(assert) {
+    const mockSystem = Service.extend({
+      activeNamespace: {
+        name: 'aNamespace',
+      },
+    });
+
+    const mockToken = Service.extend({
+      selfToken: { type: 'client' },
+      selfTokenPolicies: [
+        {
+          rulesJSON: {
+            Namespaces: [
+              {
+                Name: 'aNamespace',
+                Capabilities: ['list-jobs'],
+              },
+            ],
+          },
+        },
+      ],
+    });
+
+    this.owner.register('service:system', mockSystem);
+    this.owner.register('service:token', mockToken);
+
+    const jobAbility = this.owner.lookup('ability:job');
+    assert.notOk(jobAbility.canRun);
+  });
+
+  test('it handles globs in namespace names', function(assert) {
+    const mockSystem = Service.extend({
+      activeNamespace: {
+        name: 'aNamespace',
+      },
+    });
+
+    const mockToken = Service.extend({
+      selfToken: { type: 'client' },
+      selfTokenPolicies: [
+        {
+          rulesJSON: {
+            Namespaces: [
+              {
+                Name: 'production-*',
+                Capabilities: ['submit-job'],
+              },
+              {
+                Name: 'production-api',
+                Capabilities: ['submit-job'],
+              },
+              {
+                Name: 'production-web',
+                Capabilities: [],
+              },
+              {
+                Name: '*-suffixed',
+                Capabilities: ['submit-job'],
+              },
+              {
+                Name: '*-more-suffixed',
+                Capabilities: [],
+              },
+              {
+                Name: '*-abc-*',
+                Capabilities: ['submit-job'],
+              },
+            ],
+          },
+        },
+      ],
+    });
+
+    this.owner.register('service:system', mockSystem);
+    this.owner.register('service:token', mockToken);
+
+    const jobAbility = this.owner.lookup('ability:job');
+    const systemService = this.owner.lookup('service:system');
+
+    systemService.set('activeNamespace.name', 'production-web');
+    assert.notOk(jobAbility.canRun);
+
+    systemService.set('activeNamespace.name', 'production-api');
+    assert.ok(jobAbility.canRun);
+
+    systemService.set('activeNamespace.name', 'production-other');
+    assert.ok(jobAbility.canRun);
+
+    systemService.set('activeNamespace.name', 'something-suffixed');
+    assert.ok(jobAbility.canRun);
+
+    systemService.set('activeNamespace.name', 'something-more-suffixed');
+    assert.notOk(
+      jobAbility.canRun,
+      'expected the namespace with the greatest number of matched characters to be chosen'
+    );
+
+    systemService.set('activeNamespace.name', '000-abc-999');
+    assert.ok(jobAbility.canRun, 'expected to be able to match against more than one wildcard');
+  });
+});
```

Comment:
Testing taxonomy/nomenclature is always subjective but to me this does qualify as a unit test because I’m mocking the interfaces with the rest of the application that the unit is interacting with; for me, an integration test wouldn’t use mocking. I don’t tend to extract things like the namespace-matching until they’re useful elsewhere in the application, which this likely will never be, unless I need to for testing reasons, which I don’t think I do in this case. But I recognise this is a realm of many opinions and little consensus 🤓

Reply:
Cool, thanks for the thoughts!